tensorflow_datasets/core/download/kaggle.py

```diff
@@ -60,19 +62,24 @@ def filename(self):
     return self._filename
 
   @classmethod
-  def from_url(cls, url):
+  def from_url(cls, url, download_type):
     if not KaggleFile.is_kaggle_url(url):
       raise TypeError("Not a valid kaggle URL")
     competition_name, filename = url[len(cls._URL_PREFIX):].split("/", 1)
+    if download_type == KaggleFile._DATASET_SUFFIX:
+      dataset_name, filename = filename.split("/", 1)
+      competition_name = "%s/%s" % (competition_name, dataset_name)
     return cls(competition_name, filename)
 
   @staticmethod
   def is_kaggle_url(url):
     return url.startswith(KaggleFile._URL_PREFIX)
 
   def to_url(self):
-    return "%s%s/%s" % (self._URL_PREFIX, self._competition_name,
-                        self._filename)
+    dataset_type = int("/" in self._competition_name)
+    return "%s%s/%s%d" % (self._URL_PREFIX, self._competition_name,
+                          self._filename,
+                          KaggleFile._DATASET_COMPETITION_SUFFIX[dataset_type])
```

Comment:
If I understand correctly, you are hard-coding a custom suffix to the url, then strip it when registering the checksum.

This looks quite hacky and error prone to me. Could you give more context why this is required and why can't the dataset be detected directly in `KaggleCompetitionDownloader.download_file` ?

Reply:
The main reason for this fix is, the downloader doesn't support downloading Kaggle datasets file. The main reason being, `download_file()` can only download from kaggle competitions. (See the command, line 131 - 141, that is being run. It is of the form `kaggle competitions . . .`). To add support to downloading datasets from kaggle, `kaggle datasets . . .` must be run.
So, the idea is adding a suffix to the URL which denotes if the link is a dataset link or competition link or not and then processing the splitting on line 82

Reply:
What I don't understand is why can't you detect the type from the url directly in the `from_url()` function, without all the suffix forwarding which feel hacky ?

Reply:
The main reason is, inside from_url() we are getting the complete "kaggle://" url and no other information.
So, it's hard to differentiate between competition URLs and Dataset URLs.
For example the URL may look like:
kaggle://competition-name/foldername/filename.csv

Or,
kaggle://dataset-name/author/dataset.csv

In such case, upon splitting the url in Competition-name and filename, or, dataset-name and filename. It might cause some trouble.
For the firstcase, the splitting should be done as
Competition-name"

Reply:
@Conchylicultor

Reply:
In this case, could you instead modify the prefix ? I think it would be more readable:

```
dl_manager.download('kaggle-dataset://dataset-name/author/dataset.csv')
dl_manager.download('kaggle-competition://dataset-name/author/dataset.csv')
```

Reply:
@Conchylicultor , Instead of that, I added `kaggle://d/author/dataset.csv` `kaggle://c/competition-name/foldername/filename.csv`
Do you think that would work? If you think that's a bad idea, I can make it the way you asked for.
The main motivation of choosing `kaggle://c/` and `kaggle://d/` was to make the link more structural, and something similar to what kaggle links actually look like.

Reply:
You're right. That's better to have a 2-level structure.
But could you update to `kaggle://competition/...` and `kaggle://dataset/...`. I think a single letter would be confusing for new users reading the code.

Reply:
@Conchylicultor Pushed requested changes. :)